matroid/src/helpers.py

```
@@ -0,0 +1,204 @@
+import datetime
+import os
+import requests
+
+from matroid import error
+
+MAX_LOCAL_IMAGE_SIZE = 50 * 1024 * 1024
+MAX_LOCAL_IMAGE_BATCH_SIZE = 50 * 1024 * 1024
+
+
+def api_call(default_error):
+    """setup and teardown decorator for API calls"""
+    def decorator(func):
+        def setup_and_teardown(self, *original_args, **original_kwargs):
+            self.retrieve_token()
+
+            response = func(self, *original_args, **original_kwargs)
+
+            try:
+                self.check_errors(response, default_error)
+            except error.TokenExpirationError:
+                self.retrieve_token(options={'request_from_server': True})
+                return func(self, *original_args, **original_kwargs)
+            else:
+                return self.format_response(response)
+        return setup_and_teardown
+    return decorator
+
+
+def bytes_to_mb(self, bytes):
+    return bytes / 1024 / 1024
+
+
+def check_errors(self, response=None, UserErr=None):
+    """Raise specific errors depending on how the API call failed"""
+    status = response.status_code
+    code = None
+    try:
+        code = response.json().get('code')
+    except:
+        pass
+
+    if status == 429 and code == 'rate_err':
+        raise error.RateLimitError(response)
+    elif status == 402 and code == 'payment_err':
+        raise error.PaymentError(response)
+    elif status / 100 == 4:
+        if code == 'token_expiration_err':
+            raise error.TokenExpirationError(response)
+        elif UserErr:
+            raise UserErr(response)
+        else:
+            raise error.APIError(response)
+    elif code == 'media_err':
+        raise error.MediaError(response)
+    elif status / 100 == 5 and code == 'server_err':
+        raise error.ServerError(response)
+    elif status / 100 != 2:
+        raise error.APIError(response)
```

Can there be any other error codes? Might it end up just dropping through and failing silently?

---

This covers 2xx, 4xx, and 5xx, which seems to be enough for most cases, maybe I can add an `OtherError` or `UnknownError` to catch all other errors, and if we get that type of errors we can add specific one for them later?

---

Do we get other errors if the user tries to use http instead of https?

---

I tried sending requests of `get account info/create stream/create monitoring/get monitoring result/kill monitoring/delete monitoring` to url `http://www.matroid.com/api/v1` and `http://staging.dev.matroid.com/api/v1`, they seem to work just the same🤔

---

probably getting redirected then. That's fine.